File: packages/auth-client/src/python/aeos_auth_client/rbac.py
from __future__ import annotations

import os
import httpx
from fastapi import HTTPException, status

from .types import AuthContext, PermissionCheckResult


async def check_permission(ctx: AuthContext, resource: str, action: str) -> PermissionCheckResult:
    auth_service_url = os.environ.get("AUTH_SERVICE_URL")
    if not auth_service_url:
        raise RuntimeError("AUTH_SERVICE_URL not configured")

    async with httpx.AsyncClient() as client:
        response = await client.post(
            f"{auth_service_url}/v1/rbac/check",
            json={"user_id": ctx.user_id, "tenant_id": ctx.tenant_id, "resource": resource, "action": action},
        )
    if not response.is_success:
        raise RuntimeError(f"RBAC check failed: {response.status_code}")

    body = response.json()
    return PermissionCheckResult(allowed=body["allowed"], reason=body.get("reason"))


async def require_permission(ctx: AuthContext, resource: str, action: str) -> None:
    result = await check_permission(ctx, resource, action)
    if not result.allowed:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"Forbidden: {resource}:{action}",
        )

File: packages/auth-client/src/rbac.ts
import type { AuthContext, PermissionCheckResult } from './types.js';

export async function checkPermission(
  ctx: AuthContext,
  resource: string,
  action: string,
): Promise<PermissionCheckResult> {
  const authServiceUrl = process.env['AUTH_SERVICE_URL'];
  if (!authServiceUrl) {
    throw new Error('AUTH_SERVICE_URL not configured');
  }

  const response = await fetch(`${authServiceUrl}/v1/rbac/check`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      user_id: ctx.userId,
      tenant_id: ctx.tenantId,
      resource,
      action,
    }),
  });

  if (!response.ok) {
    throw new Error(`RBAC check failed: ${response.status}`);
  }

  const body = (await response.json()) as { allowed: boolean; reason?: string };
  return { allowed: body.allowed, reason: body.reason };
}

export async function requirePermission(
  ctx: AuthContext,
  resource: string,
  action: string,
): Promise<void> {
  const result = await checkPermission(ctx, resource, action);
  if (!result.allowed) {
    const err = new Error(`Forbidden: ${resource}:${action}`);
    (err as NodeJS.ErrnoException).code = 'FORBIDDEN';
    throw err;
  }
}
